File: shared/python/error_handler.py
"""
Error handling utilities and decorators for DGMSTT Python projects
"""

import functools
import logging
import traceback
import time
from typing import (
    Any, Callable, Optional, Type, TypeVar, Union, Tuple, Dict, List,
    Protocol, runtime_checkable
)
from datetime import datetime
import json

T = TypeVar('T')
F = TypeVar('F', bound=Callable[..., Any])

# Set up module logger
logger = logging.getLogger(__name__)


class BaseError(Exception):
    """Base error class with structured data support"""
    
    def __init__(
        self,
        message: str,
        code: Optional[str] = None,
        data: Optional[Dict[str, Any]] = None,
        cause: Optional[Exception] = None
    ):
        super().__init__(message)
        self.message = message
        self.code = code or self.__class__.__name__
        self.data = data or {}
        self.cause = cause
        self.timestamp = datetime.now()
        
    def to_dict(self) -> Dict[str, Any]:
        """Convert error to dictionary representation"""
        result = {
            'type': self.__class__.__name__,
            'message': self.message,
            'code': self.code,
            'timestamp': self.timestamp.isoformat(),
            'data': self.data
        }
        if self.cause:
            result['cause'] = str(self.cause)
        return result
        
    def to_json(self) -> str:
        """Convert error to JSON string"""
        return json.dumps(self.to_dict())


class ValidationError(BaseError):
    """Raised when validation fails"""
    def __init__(self, message: str, field: Optional[str] = None, value: Any = None):
        data = {}
        if field:
            data['field'] = field
        if value is not None:
            data['value'] = str(value)
        super().__init__(message, code='VALIDATION_ERROR', data=data)


class NetworkError(BaseError):
    """Raised when network operations fail"""
    def __init__(
        self,
        message: str,
        url: Optional[str] = None,
        status_code: Optional[int] = None
    ):
        data = {}
        if url:
            data['url'] = url
        if status_code:
            data['status_code'] = status_code
        super().__init__(message, code='NETWORK_ERROR', data=data)


class TimeoutError(BaseError):
    """Raised when operations timeout"""
    def __init__(self, message: str, operation: str, timeout_seconds: float):
        super().__init__(
            message,
            code='TIMEOUT_ERROR',
            data={'operation': operation, 'timeout_seconds': timeout_seconds}
        )


@runtime_checkable
class Retriable(Protocol):
    """Protocol for retriable exceptions"""
    @property
    def retriable(self) -> bool:
        """Whether the error is retriable"""
        ...


def retry(
    max_attempts: int = 3,
    delay: float = 1.0,
    backoff: float = 2.0,
    exceptions: Tuple[Type[Exception], ...] = (Exception,),
    on_retry: Optional[Callable[[Exception, int], None]] = None,
    should_retry: Optional[Callable[[Exception], bool]] = None
) -> Callable[[F], F]:
    """
    Decorator for retrying functions with exponential backoff
    
    Args:
        max_attempts: Maximum number of attempts
        delay: Initial delay between retries in seconds
        backoff: Backoff multiplier
        exceptions: Tuple of exceptions to catch
        on_retry: Callback called on each retry with (exception, attempt)
        should_retry: Function to determine if an exception should trigger retry
        
    Example:
        @retry(max_attempts=3, delay=1.0)
        def fetch_data():
            return requests.get('http://api.example.com/data')
    """
    def decorator(func: F) -> F:
        @functools.wraps(func)
        def wrapper(*args, **kwargs):
            last_exception = None
            current_delay = delay
            
            for attempt in range(1, max_attempts + 1):
                try:
                    return func(*args, **kwargs)
                except exceptions as e:
                    last_exception = e
                    
                    # Check if we should retry
                    if should_retry and not should_retry(e):
                        raise
                        
                    # Check if exception implements Retriable protocol
                    if isinstance(e, Retriable) and not e.retriable:
                        raise
                        
                    if attempt == max_attempts:
                        raise
                        
                    # Call retry callback
                    if on_retry:
                        on_retry(e, attempt)
                    else:
                        logger.warning(
                            f"Retry {attempt}/{max_attempts} for {func.__name__} "
                            f"after {e.__class__.__name__}: {str(e)}"
                        )
                    
                    # Wait before retry
                    time.sleep(current_delay)
                    current_delay *= backoff
                    
            # This should never be reached, but just in case
            if last_exception:
                raise last_exception
                
        return wrapper
    return decorator


def catch_and_log(
    exceptions: Tuple[Type[Exception], ...] = (Exception,),
    default: Any = None,
    log_level: int = logging.ERROR,
    log_traceback: bool = True,
    reraise: bool = False
) -> Callable[[F], F]:
    """
    Decorator to catch exceptions and log them
    
    Args:
        exceptions: Tuple of exceptions to catch
        default: Default value to return on exception
        log_level: Logging level for errors
        log_traceback: Whether to log the full traceback
        reraise: Whether to re-raise the exception after logging
        
    Example:
        @catch_and_log(exceptions=(ValueError,), default=0)
        def parse_int(value):
            return int(value)
    """
    def decorator(func: F) -> F:
        @functools.wraps(func)
        def wrapper(*args, **kwargs):
            try:
                return func(*args, **kwargs)
            except exceptions as e:
                message = f"Error in {func.__name__}: {e.__class__.__name__}: {str(e)}"
                
                if log_traceback:
                    message += f"\n{traceback.format_exc()}"
                    
                logger.log(log_level, message)
                
                if reraise:
                    raise
                    
                return default
                
        return wrapper
    return decorator


def with_timeout(timeout_seconds: float) -> Callable[[F], F]:
    """
    Decorator to add timeout to functions (requires signal support)
    
    Note: This only works on Unix-like systems and in the main thread
    
    Args:
        timeout_seconds: Timeout in seconds
        
    Example:
        @with_timeout(5.0)
        def slow_operation():
            time.sleep(10)
    """
    def decorator(func: F) -> F:
        @functools.wraps(func)
        def wrapper(*args, **kwargs):
            import signal
            
            def timeout_handler(signum, frame):
                raise TimeoutError(
                    f"Operation timed out after {timeout_seconds} seconds",
                    operation=func.__name__,
                    timeout_seconds=timeout_seconds
                )
            
            # Set up timeout
            old_handler = signal.signal(signal.SIGALRM, timeout_handler)
            signal.alarm(int(timeout_seconds))
            
            try:
                result = func(*args, **kwargs)
            finally:
                # Clean up
                signal.alarm(0)
                signal.signal(signal.SIGALRM, old_handler)
                
            return result
            
        return wrapper
    return decorator


def handle_errors(
    error_map: Dict[Type[Exception], Union[Any, Callable[[Exception], Any]]]
) -> Callable[[F], F]:
    """
    Decorator to handle specific exceptions with custom handlers
    
    Args:
        error_map: Dictionary mapping exception types to handlers or values
        
    Example:
        @handle_errors({
            ValueError: 0,
            KeyError: lambda e: f"Missing key: {e}",
            IOError: lambda e: None
        })
        def process_data(data):
            return data['value']
    """
    def decorator(func: F) -> F:
        @functools.wraps(func)
        def wrapper(*args, **kwargs):
            try:
                return func(*args, **kwargs)
            except Exception as e:
                for exc_type, handler in error_map.items():
                    if isinstance(e, exc_type):
                        if callable(handler):
                            return handler(e)
                        else:
                            return handler
                # Re-raise if no handler found
                raise
                
        return wrapper
    return decorator


class ErrorContext:
    """
    Context manager for adding context to errors
    
    Example:
        with ErrorContext("Processing user data", user_id=123):
            process_user(123)
    """
    
    def __init__(self, operation: str, **context):
        self.operation = operation
        self.context = context
        
    def __enter__(self):
        return self
        
    def __exit__(self, exc_type, exc_val, exc_tb):
        if exc_val is not None:
            # Add context to the exception
            if isinstance(exc_val, BaseError):
                exc_val.data.update(self.context)
                exc_val.data['operation'] = self.operation
            else:
                # Wrap in BaseError with context
                wrapped = BaseError(
                    f"{self.operation} failed: {str(exc_val)}",
                    code=exc_type.__name__,
                    data=self.context,
                    cause=exc_val
                )
                raise wrapped from exc_val
        return False


def safe_execute(
    func: Callable[..., T],
    *args,
    default: Optional[T] = None,
    exceptions: Tuple[Type[Exception], ...] = (Exception,),
    **kwargs
) -> Tuple[Optional[T], Optional[Exception]]:
    """
    Safely execute a function and return result with error
    
    Args:
        func: Function to execute
        *args: Positional arguments for func
        default: Default value on error
        exceptions: Exceptions to catch
        **kwargs: Keyword arguments for func
        
    Returns:
        Tuple of (result, error)
        
    Example:
        result, error = safe_execute(int, "123")
        if error:
            print(f"Conversion failed: {error}")
    """
    try:
        result = func(*args, **kwargs)
        return result, None
    except exceptions as e:
        return default, e


def create_error_handler(
    default_response: Any = None,
    log_errors: bool = True,
    include_traceback: bool = False
) -> Callable[[Callable], Callable]:
    """
    Create a custom error handler decorator
    
    Args:
        default_response: Default response on error
        log_errors: Whether to log errors
        include_traceback: Whether to include traceback in response
        
    Returns:
        Error handler decorator
    """
    def error_handler(func: Callable) -> Callable:
        @functools.wraps(func)
        def wrapper(*args, **kwargs):
            try:
                return func(*args, **kwargs)
            except Exception as e:
                if log_errors:
                    logger.error(
                        f"Error in {func.__name__}: {e}",
                        exc_info=include_traceback
                    )
                    
                if include_traceback:
                    error_info = {
                        'error': str(e),
                        'type': e.__class__.__name__,
                        'traceback': traceback.format_exc()
                    }
                else:
                    error_info = {
                        'error': str(e),
                        'type': e.__class__.__name__
                    }
                    
                if callable(default_response):
                    return default_response(error_info)
                else:
                    return default_response
                    
        return wrapper
    return error_handler


class ErrorCollector:
    """
    Collect multiple errors and handle them together
    
    Example:
        collector = ErrorCollector()
        
        with collector.collect():
            validate_name(data['name'])
            
        with collector.collect():
            validate_email(data['email'])
            
        if collector.has_errors():
            raise ValidationError(collector.format_errors())
    """
    
    def __init__(self):
        self.errors: List[Tuple[str, Exception]] = []
        
    def collect(self, context: str = ""):
        """Context manager to collect errors"""
        return ErrorCollectContext(self, context)
        
    def add_error(self, error: Exception, context: str = ""):
        """Add an error to the collection"""
        self.errors.append((context, error))
        
    def has_errors(self) -> bool:
        """Check if any errors were collected"""
        return len(self.errors) > 0
        
    def format_errors(self) -> str:
        """Format all errors as a string"""
        if not self.errors:
            return ""
            
        lines = ["Multiple errors occurred:"]
        for i, (context, error) in enumerate(self.errors, 1):
            if context:
                lines.append(f"{i}. [{context}] {error}")
            else:
                lines.append(f"{i}. {error}")
                
        return "\n".join(lines)
        
    def get_errors(self) -> List[Tuple[str, Exception]]:
        """Get all collected errors"""
        return self.errors.copy()
        
    def clear(self):
        """Clear all collected errors"""
        self.errors.clear()


class ErrorCollectContext:
    """Context manager for ErrorCollector"""
    
    def __init__(self, collector: ErrorCollector, context: str):
        self.collector = collector
        self.context = context
        
    def __enter__(self):
        return self
        
    def __exit__(self, exc_type, exc_val, exc_tb):
        if exc_val is not None:
            self.collector.add_error(exc_val, self.context)
            return True  # Suppress the exception
        return False

File: shared/typescript/error-handler.ts
/**
 * Common error handling utilities for DGMSTT TypeScript projects
 */

import { z, ZodError, ZodSchema } from 'zod';

/**
 * Base error class with structured data support
 */
export abstract class BaseError extends Error {
  abstract readonly code: string;
  abstract toJSON(): Record<string, unknown>;
}

/**
 * Create a typed error class with validation
 */
export function createErrorClass<T extends ZodSchema>(
  name: string,
  code: string,
  schema: T
) {
  return class extends BaseError {
    readonly code = code;
    
    constructor(
      public readonly data: z.input<T>,
      options?: ErrorOptions
    ) {
      super(`${name}: ${JSON.stringify(data)}`, options);
      this.name = name;
      
      // Validate data
      try {
        schema.parse(data);
      } catch (e) {
        throw new Error(`Invalid error data for ${name}: ${e}`);
      }
    }
    
    toJSON() {
      return {
        name: this.name,
        code: this.code,
        message: this.message,
        data: this.data,
      };
    }
    
    static isInstance(error: unknown): error is InstanceType<ReturnType<typeof createErrorClass>> {
      return error instanceof Error && error.name === name;
    }
  };
}

/**
 * Common error types
 */
export const ValidationError = createErrorClass(
  'ValidationError',
  'VALIDATION_ERROR',
  z.object({
    field: z.string(),
    message: z.string(),
    value: z.unknown().optional(),
  })
);

export const NetworkError = createErrorClass(
  'NetworkError',
  'NETWORK_ERROR',
  z.object({
    url: z.string(),
    method: z.string().optional(),
    statusCode: z.number().optional(),
    message: z.string(),
  })
);

export const TimeoutError = createErrorClass(
  'TimeoutError',
  'TIMEOUT_ERROR',
  z.object({
    operation: z.string(),
    timeoutMs: z.number(),
  })
);

/**
 * Safely execute an async operation with error handling
 */
export async function safeAsync<T>(
  operation: () => Promise<T>,
  fallback?: T
): Promise<{ data?: T; error?: Error }> {
  try {
    const data = await operation();
    return { data };
  } catch (error) {
    if (fallback !== undefined) {
      return { data: fallback, error: error as Error };
    }
    return { error: error as Error };
  }
}

/**
 * Retry an operation with exponential backoff
 */
export async function retryWithBackoff<T>(
  operation: () => Promise<T>,
  options: {
    maxAttempts?: number;
    initialDelayMs?: number;
    maxDelayMs?: number;
    shouldRetry?: (error: Error) => boolean;
  } = {}
): Promise<T> {
  const {
    maxAttempts = 3,
    initialDelayMs = 1000,
    maxDelayMs = 30000,
    shouldRetry = () => true,
  } = options;

  let lastError: Error;
  
  for (let attempt = 0; attempt < maxAttempts; attempt++) {
    try {
      return await operation();
    } catch (error) {
      lastError = error as Error;
      
      if (attempt === maxAttempts - 1 || !shouldRetry(lastError)) {
        throw lastError;
      }
      
      const delay = Math.min(
        initialDelayMs * Math.pow(2, attempt),
        maxDelayMs
      );
      
      await new Promise(resolve => setTimeout(resolve, delay));
    }
  }
  
  throw lastError!;
}

/**
 * Type guard for checking if an error has a specific code
 */
export function hasErrorCode(
  error: unknown,
  code: string
): error is Error & { code: string } {
  return (
    error instanceof Error &&
    'code' in error &&
    (error as any).code === code
  );
}

/**
 * Convert unknown error to Error instance
 */
export function toError(error: unknown): Error {
  if (error instanceof Error) {
    return error;
  }
  
  if (typeof error === 'string') {
    return new Error(error);
  }
  
  if (error && typeof error === 'object' && 'message' in error) {
    return new Error(String(error.message));
  }
  
  return new Error(String(error));
}

/**
 * Create a detailed error message with context
 */
export function createErrorMessage(
  operation: string,
  error: Error,
  context?: Record<string, unknown>
): string {
  const parts = [`${operation} failed: ${error.message}`];
  
  if (context && Object.keys(context).length > 0) {
    parts.push(`Context: ${JSON.stringify(context)}`);
  }
  
  if (error.stack) {
    parts.push(`Stack: ${error.stack}`);
  }
  
  return parts.join('\n');
}

/**
 * Error handler for JSON-RPC style errors
 */
export interface JsonRpcError {
  code: number;
  message: string;
  data?: unknown;
}

export class JsonRpcErrorResponse extends Error {
  constructor(
    public readonly error: JsonRpcError,
    public readonly id: string | number | null
  ) {
    super(error.message);
    this.name = 'JsonRpcError';
  }
  
  toJSON() {
    return {
      jsonrpc: '2.0',
      error: this.error,
      id: this.id,
    };
  }
}

/**
 * Standard JSON-RPC error codes
 */
export const JsonRpcErrorCode = {
  ParseError: -32700,
  InvalidRequest: -32600,
  MethodNotFound: -32601,
  InvalidParams: -32602,
  InternalError: -32603,
  ServerError: -32000,
} as const;

/**
 * Create standard JSON-RPC errors
 */
export const createJsonRpcError = {
  parseError: (id: string | number | null, data?: unknown) =>
    new JsonRpcErrorResponse(
      { code: JsonRpcErrorCode.ParseError, message: 'Parse error', data },
      id
    ),
    
  invalidRequest: (id: string | number | null, data?: unknown) =>
    new JsonRpcErrorResponse(
      { code: JsonRpcErrorCode.InvalidRequest, message: 'Invalid Request', data },
      id
    ),
    
  methodNotFound: (id: string | number | null, method?: string) =>
    new JsonRpcErrorResponse(
      { 
        code: JsonRpcErrorCode.MethodNotFound, 
        message: `Method not found${method ? `: ${method}` : ''}` 
      },
      id
    ),
    
  invalidParams: (id: string | number | null, data?: unknown) =>
    new JsonRpcErrorResponse(
      { code: JsonRpcErrorCode.InvalidParams, message: 'Invalid params', data },
      id
    ),
    
  internalError: (id: string | number | null, data?: unknown) =>
    new JsonRpcErrorResponse(
      { code: JsonRpcErrorCode.InternalError, message: 'Internal error', data },
      id
    ),
};